src/Client.php

```diff
@@ -28,60 +36,78 @@ class Client
      */
     protected $httpClient;
 
+    /**
+     * @var StreamFactoryInterface
+     */
+    protected $streamFactory;
+
+    /**
+     * @var RequestFactoryInterface
+     */
+    protected $requestFactory;
+
     /**
      * @var array
      */
     protected $httpHeaders;
 
+    /**
+     * @var array
+     */
+    protected $httpOptions;
+
```

Comment:
I think that a lot of what the PSR-18 interface did was take away the more granular configuration options. This is one of the only ones that I saw left, so I tried to keep it.

In another place it also pulls the value for the `headers` key out of the `httpOptions` and merges them with `authorizationHeaders` to get the headers variable, so my decision to keep it was to make less breaking changes in the new version of this library. But I agree that we could just have a `headers` variable (and a `version` variable if we still want to support the version).

I think that the idea behind PSR-18 to strip away some of the other configuration options was on purpose. I think they wanted library creators to let the end user pass them a client with certain configuration if they wanted that configuration, that way each client could still have it's own configuration settings but be PSR-18 compliant. I don't think there's any way to configure the PSR-18 client beyond the HTTP version. I might have misunderstood this aspect of it when reading about it though so I could be wrong.

Reply:
I think you got this correct, what you're saying makes sense to me.

The way this was implemented in the current version was really scrappy and without much thought behind it. I started by supporting passing only headers, so I had the headers argument, then people started asking for ability to configure more options in the request, like timeout, redirects, etc, so I provided the ability to pass the whole `httpOptions` array, which is a guzzle specific thing. Because `httpOptions` had the headers key too, it became redundant, but I couldn't remove the headers passed separately because this would break backwards compatibility.

That's why I wouldn't worry too much about sticking to the current design, but rather to try to reach a design that's future proof.

Because of the way this package works, where we construct requests on behalf of users and send them to graphQL servers, we're restricted when it comes to forming and manipulating HTTP requests.

When relying on guzzle, it was simple to just let users pass the `httpOptions`, and guzzle would do the request on our behalf for them. Now that we're moving towards a design that allows for any PSR-18 compliant http client, it becomes tricky to handle the request manipulation.

Because the `RequestInterface` is minimal and doesn't provide methods to manipulate other request options, we have 2 ways to try to approach this to provide our users with flexibility in constructing their requests:
1. Give them the ability to inject their own `RequestDecorator`  into the client: This would allow them to manipulate their requests without making any changes to the client logic. We would need to create a `RequestDecoratorInterface` that they can implement, or they can extend the existing `RequestDecorator`.
2. Add an extra parameter to the send request methods in the client that would have a nullable `RequestInterface` object: his would allow the users to provide us with a request object that we can manipulate to add the body to it before sending it to the API server. I don't like this option though, it seems dirty to me.

Besides those two approaches, I would like to preserve the ability to use guzzle to create http requests with guzzle options, as I imagine a lot of people would find it hard to manipulate the request objects themselves. We can have a new method in the GraphQlClient, something like `runQueryWithGuzzle` that would use `modify_request` from https://github.com/guzzle/psr7#function-modify_request to create http requests for users easily. We would want to composer suggest `guzzlehttp/psr7` but _not_ require it in the package and throw exception if the method is called and the function doesn't exist.

I think that this solution would preserve the ease of creating API requests and at the same time not force users to stick to one http client implementation. If they do not want to use guzzle that would be fine, but they will have to make their own request manipulation if they want to set advanced request options.

Let me know if you think this makes sense and if you have any better ideas.

Reply:
Hmm, a lot to think about here.

So with the current implementation, users can pass in their own preconfigured PSR-18 client when building the GraphQL client. This doesn't allow manipulation of the actual request but it does allow for them to set http options. So instead of a `runQueryWithGuzzle` method, the user could make a new Guzzle client with certain http options and pass it into the GraphQL client constructor. This would work with any http client the user use. This might be sufficient enough for some users we can document it well, but I could see how there could be some instances where the user would also need to modify the request itself. 

As far as injecting their own RequestDecorator, this seems like it could be a good idea but I'm wondering if it would be confusing for some users to understand what they're supposed to do. It would definitely follow the composition over inheritance principle, but it might be confusing to users that they have to pass a class in which would modify the request, rather than modifying a request directly. I'm not sure though, do we know of any other libraries which do something like this? I would be interested to see in general.

It might make sense to have two separate ways to make GraphQL requests, one being the normal way and the other which would let the user access the request before it is sent to the server. For example...

```php
// actual code would look different
public function runQuery($query, bool $resultsAsArray = false, array $variables = []): Results
{
    $request = $this->createRequest($query, $resultsAsArray, $variables);
    $response = $this->sendRequest($request);
    ...
}

public function createRequest($query, bool $resultsAsArray = false, array $variables = []): RequestInterface
{
    ...
}

public function sendRequest(RequestInterface $request): Results
{
   ...
}
```

This would allow the user to either use the default method `runQuery` which would just send the request, or they could take advantage of the new `createRequest` method if they wanted to get a request back that they could modify before passing it back to the client to make the request. This would also allow users to use the library incorrectly and create their own request completely and send it to the library though. I can't think of a way of preventing this but I'm not sure if that's an actual problem. What do you think about this solution?

Reply:
I spent some time reading and thinking about this. To be honest I'm not a big fan of splitting the process of sending requests into two steps. It kind of defeats the purpose of having a library that simplifies the process.

I agree with you on the custom decorator solution as well, it can be confusing if not well document *and* well explained. But even then it still wouldn't be intuitive.

Inject the logic of request manipulation in the http client too seems a bit out of place to me since the client is just supposed to be responsible for sending the request.

The conclusion I reached is that the request factory should be the best place for users to manipulate their requests if they want to make custom changes to them. They can implement their own request factory that contains all the http request options they want to create the request with and we will modify this request to add the body (and possibly headers) to it.

Before I dive into the details of the implementation I just want to make sure that you're onboard with me on this conclusion. Let me know your thoughts.

Reply:
Just doing a quick recap here to make sure I didn't miss any important changes. I forgot to respond but I definitely agree with your conclusion. With the ability for them to pass in their own PSR-7 compliant Request Factory, as well as their own PSR-18 Client, I think the user will have plenty of customization available if they want to go that route. While, they're still getting the ease of just passing in some headers if they need.

This sounds exactly what PHP-FIG had in mind when they created these PSRs.